uamqp/client.py

```diff
@@ -237,6 +262,99 @@ def do_work(self, **kwargs):
         return self._client_run(**kwargs)
 
 
+class SendClient(AMQPClient):
+    def __init__(self, hostname, target, auth=None, **kwargs):
+        self.target = target
+        # Sender and Link settings
+        self._max_message_size = kwargs.pop('max_message_size', None) or _MAX_FRAME_SIZE_BYTES
+        self._link_properties = kwargs.pop('link_properties', None)
+        self._link_credit = kwargs.pop('link_credit', None)
+        super(SendClient, self).__init__(hostname, auth=auth, **kwargs)
+
+    def _client_ready(self):
+        """Determine whether the client is ready to start receiving messages.
+        To be ready, the connection must be open and authentication complete,
+        The Session, Link and MessageReceiver must be open and in non-errored
+        states.
+
+        :rtype: bool
+        :raises: ~uamqp.errors.MessageHandlerError if the MessageReceiver
+         goes into an error state.
+        """
+        # pylint: disable=protected-access
+        if not self._link:
+            self._link = self._session.create_sender_link(
+                target_address=self.target,
+                link_credit=self._link_credit,
+                send_settle_mode=self._send_settle_mode,
+                rcv_settle_mode=self._receive_settle_mode,
+                max_message_size=self._max_message_size,
+                properties=self._link_properties)
+            self._link.attach()
+            return False
+        if self._link.state.value != 3:  # ATTACHED
+            return False
+        return True
+
+    def _client_run(self, **kwargs):
+        """MessageSender Link is now open - perform message send
+        on all pending messages.
+        Will return True if operation successful and client can remain open for
+        further work.
+
+        :rtype: bool
+        """
+        try:
+            self._connection.listen(wait=self._socket_timeout, **kwargs)
+        except ValueError:
+            _logger.info("Timeout reached, closing sender.")
+            self._shutdown = True
+            return False
+        return True
+
+    def _transfer_message(self, message_delivery, timeout=0):
+        message_delivery.state = MessageDeliveryState.WaitingForSendAck
+        on_send_complete = partial(self._on_send_complete, message_delivery)
+        delivery = self._link.send_transfer(
+            message_delivery.message,
+            on_send_complete=on_send_complete,
+            timeout=timeout
+        )
+        if not delivery.sent:
+            raise RuntimeError("Message is not sent.")
+
+    def _on_send_complete(self, message_delivery, message, reason, state):
```

Comment:
If I recall correctly - this callback can also be called if the message expires or the link goes down. Though we can probably add handling for that as needed.

Reply:
added a "TODO" comment in the callback so that we don't forget to revisit this part in the future